### PR TITLE
RUMM-721 First party resources

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Adds a `first_party` attribute on RUM Resource to categorise them properly

### Additional Notes

This PR also refactors a couple of class to share the same detection for Traces and RUM

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [x] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [x] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt [hunk 1]
```diff
@@ -82,11 +84,22 @@ internal constructor(
...
 
+    /**
+     * Creates a [TracingInterceptor] to automatically create a trace around OkHttp [Request]s.
+     *
+     * @param tracedRequestListener a listener for automatically created [Span]s
+     */
+    @JvmOverloads
+    constructor(
+        tracedRequestListener: TracedRequestListener = NoOpTracedRequestListener()
+    ) : this(emptyList(), tracedRequestListener, { AndroidTracer.Builder().build() })
```
Comment (ncreated) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt:101]: As this configuration is now less obvious (previously the hosts were expected in Interceptor's initializer, now they are part of the `Datadog.Config`), maybe it's worth printing console warning from here on misconfiguration? I mean: when the interceptor is created, but the `setFirstPartyHosts()` was not called.

Comment (xgouchet) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt:101]: Indeed that's a good call.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/net/FirstPartyHostDetectorTest.kt [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+        // When
+        val result = testedDetector.isFirstPartyUrl(url)
+
+        // Then
+        assertThat(result).isTrue()
+    }
+
+    @Test
+    fun `𝕄 return false 𝕎 isFirstParty(HttpUrl) {unknown host postfixed with valid host}`(
```
Comment (mariusc83) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/core/internal/net/FirstPartyHostDetectorTest.kt:98]: Don't you need one also for a broken URL ?

File: dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt [hunk 3]
```diff
@@ -130,8 +143,9 @@ internal constructor(
     // region Internal
 
     private fun isRequestTraceable(request: Request): Boolean {
-        val host = request.url().host()
-        return host.matches(hostRegex)
+        val url = request.url()
+        return CoreFeature.firstPartyHostDetector.isFirstPartyUrl(url) ||
```
Comment (mariusc83) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt:154]: I would rather inject the `firstPartyHostDetector` in the constructor. Not a big fan of having this static reference inside this function here.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt [hunk 4]
```diff
@@ -224,6 +258,7 @@ internal open class TracingInterceptorTest {
         whenever(parentSpan.context()) doReturn parentSpanContext
         whenever(mockSpanBuilder.asChildOf(parentSpanContext)) doReturn mockSpanBuilder
         fakeRequest = forgeRequest(forge) { it.tag(Span::class.java, parentSpan) }
+        doAnswer { true }.whenever(mockDetector).isFirstPartyUrl(HttpUrl.get(fakeUrl))
```
Comment (mariusc83) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt:261]: wasn't this already set up in the `set up` method ?

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt [hunk 5]
```diff
@@ -253,6 +288,7 @@ internal open class TracingInterceptorTest {
         whenever(mockTracer.extract<TextMapExtract>(any(), any())) doReturn parentSpanContext
         whenever(mockSpanBuilder.asChildOf(any<SpanContext>())) doReturn mockSpanBuilder
         fakeRequest = forgeRequest(forge)
+        doAnswer { true }.whenever(mockDetector).isFirstPartyUrl(HttpUrl.get(fakeUrl))
```
Comment (mariusc83) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt:291]: same here ?

File: dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumResourceScope.kt [hunk 6]
```diff
@@ -127,6 +128,7 @@ internal class RumResourceScope(
 
         val finalTiming = timing
         val duration = eventTime.nanoTime - startedNanos
+        val isFirstParty = CoreFeature.firstPartyHostDetector.isFirstPartyUrl(url)
```
Comment (mariusc83) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumResourceScope.kt:132]: Same here...can't we just inject the `firstPartyHostDetector` here and keep it as a singleton in the `CoreFeature` ?

**Review by mariusc83 — CHANGES_REQUESTED** [review]

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt (hunk 4 above)
Comment (xgouchet) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt:261]: Not with this value of `fakeUrl`

File: dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt (hunk 5 above)
Comment (xgouchet) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/tracing/TracingInterceptorTest.kt:291]: Not with this value of `fakeUrl`

**Review by mariusc83 — APPROVED** [review]

**Pull request merged**